pkg/qdr/qdr.go

```diff
@@ -1,73 +1,673 @@
+	for key, v1 := range b {
+		v2, ok := a[key]
+		if !ok {
+			result.Added = append(result.Added, v1)
+		} else if !reflect.DeepEqual(v1, v2) {
+			result.Deleted = append(result.Deleted, v1.Name)
+			result.Added = append(result.Added, v1)
+		}
+	}
+	for key, v1 := range a {
+		_, ok := b[key]
+		if !ok {
+			result.Deleted = append(result.Deleted, v1.Name)
+		}
+	}
+	return result
+}
+
+func (a HttpEndpointMap) Difference(b HttpEndpointMap) HttpEndpointDifference {
+	result := HttpEndpointDifference{}
+	for key, v1 := range b {
+		v2, ok := a[key]
+		if !ok {
+			result.Added = append(result.Added, v1)
+		} else if !reflect.DeepEqual(v1, v2) {
+			result.Deleted = append(result.Deleted, v1.Name)
+			result.Added = append(result.Added, v1)
+		}
+	}
+	for key, _ := range a {
+		v2, ok := b[key]
+		if !ok {
+			result.Deleted = append(result.Deleted, v2.Name)
+		}
+	}
+	return result
+}
+
+func (a *BridgeConfig) Difference(b *BridgeConfig) *BridgeConfigDifference {
+	result := BridgeConfigDifference{
+		TcpConnectors:  a.TcpConnectors.Difference(b.TcpConnectors),
+		TcpListeners:   a.TcpListeners.Difference(b.TcpListeners),
+		HttpConnectors: a.HttpConnectors.Difference(b.HttpConnectors),
+		HttpListeners:  a.HttpListeners.Difference(b.HttpListeners),
+	}
+	return &result
+}
+
+func (a *TcpEndpointDifference) Empty() bool {
+	return len(a.Deleted) == 0 && len(a.Added) == 0
+}
+
+func (a *HttpEndpointDifference) Empty() bool {
+	return len(a.Deleted) == 0 && len(a.Added) == 0
+}
+
+func (a *BridgeConfigDifference) Empty() bool {
+	return a.TcpConnectors.Empty() && a.TcpListeners.Empty() && a.HttpConnectors.Empty() && a.HttpListeners.Empty()
+}
+
+func (a *BridgeConfigDifference) Print() {
+	log.Printf("TcpConnectors added=%v, deleted=%v", a.TcpConnectors.Added, a.TcpConnectors.Deleted)
+	log.Printf("TcpListeners added=%v, deleted=%v", a.TcpListeners.Added, a.TcpListeners.Deleted)
+	log.Printf("HttpConnectors added=%v, deleted=%v", a.HttpConnectors.Added, a.HttpConnectors.Deleted)
+	log.Printf("HttpListeners added=%v, deleted=%v", a.HttpListeners.Added, a.HttpListeners.Deleted)
+}
+
+func GetRouterConfigForHeadlessProxy(definition types.ServiceInterface, siteId string, namespace string) (string, error) {
+	config := InitialConfig("$HOSTNAME", siteId, true)
+	//add edge-connector
+	config.AddSslProfile(SslProfile{
+		Name: types.InterRouterProfile,
+	})
+	config.AddConnector(Connector{
+		Name:       "uplink",
+		SslProfile: types.InterRouterProfile,
+		Host:       "skupper-internal." + namespace,
+		Port:       strconv.Itoa(int(types.EdgeListenerPort)),
+		Role:       RoleEdge,
+	})
+	config.AddListener(Listener{
+		Name: "amqp",
+		Host: "localhost",
+		Port: 5672,
+	})
+	port := definition.Port
+	if len(definition.Targets) == 1 && definition.Targets[0].TargetPort != 0 {
+		port = definition.Targets[0].TargetPort
+	}
+	if definition.Origin == "" {
+		host := definition.Headless.Name + "-${POD_ID}." + definition.Address + "." + namespace
```

Comment:
Observed issue with substitution, should this be:

host := definition.Headless.Name + "-" + "$POD_ID" + "." + definition.Address + "." + namespace

Reply:
No, I think the curly brackets should be fine... what is the sha of the router image you are using, do you know?

Reply:
Hmmm, I must have been seeing things as it is ok now. 

router:
image: quay.io/gordons/qdrouterd:0.4.0-alpha-1
imageID: quay.io/gordons/qdrouterd@sha256:45388c718efe4b5fc535f53feb543415e7261dcaa0188f5663d960b2e8c9ddd5

proxy:
image: quay.io/gordons/qdrouterd:0.4.0-alpha-1
imageID: quay.io/gordons/qdrouterd@sha256:19b9700baf9acd2fe56bd4b79ae72e1d8ac81af57ad705fb840f705db3cb1f94

Reply:
That last one seems to match the latest as reported by quay.io. The image has a small change in it that sets the POD_ID based on the HOSTNAME. Possibly the older image was used initially and so did not resolve as the env var was not set(?).

Reply:
I apologize,  I was looking at a listener, I still see the issue for the connector in the qdrouterd.json

           [
                "tcpConnector",
                {
                    "name": "egress",
                    "host": "cockroachdb-g1-${POD_ID}.cockroachdb-internal-g1.skupper",
                    "port": "26257",
                    "address": "cockroachdb-internal-g1",
                    "siteId": "c39f518f-1161-44a9-bc7a-f3009e989203"
                }
            ]

Reply:
doh! never mind me, looking at env template rather than actual /tmp/qdrouterd.json

sorry for the confusion :)